인증 로직 구현

i960107 opened the pull request:
# 🔍 해결하려는 문제
1. spring security filter chain를 통해서 oauth 로그인과 id/pw 로그인 지원.
2. oauth provider가 추가되어도 쉽게 확장 가능하도록 OCP 원칙을 따르는 설계.
3. jwt 토큰 발급시 refreshToken도 같이 발급하여서 재로그인 없이 토큰 갱신 가능하도록 지원.

## 인증과정 요약
1. id/pw나 oauth **로그인**을 하고 **토큰(access token 및 refresh token)을 발급** 받는다.
2. 클라이언트는 refreshToken을 어딘가에 저장해두고 추후 **인증이 필요한 요청시 accessToken을 헤더에 담아서 요청**한다.
4. accessToken이 만료되면 저장해두었던 **refreshToken으로 토큰 갱신**을 요청한다.
5. refreshToken이 만료되면 **재 로그인**을 한다.

# 💡핵심 변경 사항
## 1. user 테이블 구조 변경
### 1-1. oauth 가입 유저를 위한 테이블을 분리하지 않고 user테이블을 수정한 이유
- oauth 가입 유저를 위한 테이블을 **분리할 경우** 회원 조회가 필요한 모든 로직에서 1) oauth유저인지, 일반 유저인지 확인해아하거나 2) 두 개의 테이블에서 유저를 조회해야하는 등 비지니스 복잡성이 매우 증가할 것임.
- oauth 가입 유저를 위한 테이블을 **분리하지 않을 경우** 각 레코드가 불필요한 필드를 가져야해서 저장 공간이 낭비됨.
- 일반 유저와 oauth 유저를 따로 관리할 필요성이 크지 않기 때문에 저장 공간을 추가적으로 사용하더라도 기존 user테이블을 수정해서 일반 가입 유저와 oauth 유저 모두를 저장하기로 결정.
### 1-2. 수정 사항
- oauth 가입 유저와 일반 가입 유저 모두 관리할 수 있도록 수정
- password 필드 nullable 수정
- oauth_user_id, oauth_provider 필드 추가
- oauth가입 유저는 oauth_user_id, oauth_provider 필드 값 가지고 password 는 null
- 일반 가입 유저는 oauth_user_id, oauth_provider 필드 null, password는 인코딩된 패스워드.

## 2. oauth 로그인
- 엔드포인트: `/oauth2/authorization/naver` `/oauth2/authorization/kakao`
- **spring security 기본 구현 활용**: 1) oauth로그인 요청을 oauth provider의 인증 페이지로 redirect하고 2) authorization_code 방식으로 access token을 얻고 3) accessToken을 사용해 OAuth2User 객체를 얻는 과정은 spring security 제공 기능 사용.
- **AuthenticationSuccessHandler와 AuthenticationFailureHandler는 직접 구현**
- 인증 실패시 LoginFailureHandler는 401 UnAuthorized응답을 보냄. id/pw 로그인과 공통 handler 사용하여 인증 실패 로직 일관되게 유지.
- 인증 성공시 OAuth2AuthenticationSuccessHandler(extends LoginSuccessHandler)에 의해서 처리됨.
###  2-1. LoginSuccessHandler
- LoginSuccessHandler를 추상화한 이유: oauth2 로그인과 id/pw 로그인 성공시의 공통 처리 로직을 추상화해 일관된 정책(로그인 응답 형식, success handler에서 발생한 예외 처리 등) 유지 가능. 
- 공통 처리 로직: authentication으로 부터 user객체를 얻고, user객체를 활용해 jwt 토큰(access token및 refresh token)을 생성하고 응답하는 로직 처리.
- LoginSuccessHandler는 **authentication으로 부터 User객체**를 얻는 **_processAuthenticationSuccess_() 메서드를 추상화.** 
- **UserDetails와 OAuth2User 인터페이스를 구현하는 공통 객체가 아닌 User를 반환하는 이유**: 
  1) 로그인 방식에 따라서 authenticated principal이 달라짐. oauth 로그인 시에는 OAuth2User가, id/pw 로그인시에는UserDetails가 authenticated principal이 됨. 
  2) UserDetails와 OAuth2User를 모두 구현하는 객체를 구현하는 경우 처리가 복잡해지고 결합도가 올라감. DefaultOAuth2UserService를 사용할 수 없게 됨. id/pw 로그인 정책 변경으로 객체가 변경되는 경우 oauth2로그인도 영향을 받게 되어 결합도가 올라감.
  3) OAuth2User와 UserDetails 객체를 분리하고 oauth2 로그인 id/pw 로그인에서 공통으로 사용하는 user객체를 사용하는 것이 낫다고 판단.

### 2-2. OAuth2AuthenticationSuccessHandler
- oauth2 로그인 성공시 LoginSuccessHandler를 구현한 OAuth2AuthenticationSuccessHandler가 처리.
- 첫 로그인시 유저 정보 생성: oauth provider와 oauth user id 값으로 db에 저장된 유저가 없을 경우 새로 가입하는 유저로 간주 테이블에 삽입 후 반환.
- **OAuth2UserAttributeConverter**: 리소스 서버의 응답형식에 따라 OAuth2User의 attributes구조가 달라짐(키 이름). converter가 서로 다른 응답형식을 OAuth2UserInfo라는 공통된 인터페이스로 통일. converter는 enum으로 provider마다 객체를 가지고, from() 함수를 구현. OAuth2AuthenticationSuccessHandler는 OAuth2UserAttributeConverter의 static함수인 getUserInfoFromOAuth2UserAttribute()를 호출해서 registrationId로 해당하는 enum객체를 찾아서 OAuth2UserInfo를 생성 후 반환.

## 3. id/pw 로그인
- 엔드포인트 `/api/v1/login`
- 인증 실패시 LoginFailureHandler는 401 UnAuthorized응답을 보냄. oauth2 로그인과 공통 handler 사용하여 인증 실패 로직 일관되게 유지.
- 인증 성공시 UsernamePasswordAuthenticationSuccessHandler(extends LoginSuccessHandler)에 의해서 처리됨. LoginSuccessHandler의 공통 로직에 대해서는 [2-1. LoginSuccessHandler](#2-1.-loginsuccesshandler)를 참고


## 4. refreshToken 발급 로직 추가.
- refreshToken 갱신 위한 엔드포인트 추가: `/api/v1/account/refresh-token`. 인증 없이 이용가능하도록 권한 설정.

### 4-1. refreshToken 정책
- **refreshToken갱신**: refreshToken을 사용해서 **accessToken 갱신시 refreshToken도 같이 갱신**. refreshToken 만료 기간이 길어서 refreshToken 탈취 시 위험하기 때문에 refreshToken도 자주 갱신되도록 함. refreshToken의 만료기간 자체를 짧게 하는 것은 재로그인이 필요하므로 사용자 불편 초래.
- **일관된 토큰 갱신 정책 유지**: jwtService에서 createJwt(), refreshToken() 메서드만 공개해서 Jwt객체를 반환. accessToken과 refreshToken을 쌍으로 관리하고, refreshToken갱신 없이 accessToken만 갱신하는 것을 방지.
- **refreshToken 만료**: refreshToken은 다음 두가지 경우에 만료됨.
  1) refreshToken 유효기간이 지난 경우: refreshToken은 사용자가 로그인한 시점으로부터 일정 시간동안만 유효.
  2) 잘못된 refreshToken으로 토큰 갱신 요청시: userId로 저장된 refreshToken과 다른 값의 토큰으로 요청시 refreshToken이 탈취된 것으로 보고 저장된 refreshToken 삭제. 유저 재로그인 필요.
### 4-2. refreshToken을 redis에 저장하는 이유
  1. refreshToken 저장시 ttl을 설정해 일정 기간이 지나면(application.yaml에서 설정한 refreshTokenExpirySeconds) 삭제. 
  3. 기존 db에 i/o 부하 없이 빠르고 간단하게 조회 및 저장 가능.

# 🔧 추가 변경사항
## 1. 세션 관리 추가
- 세션을 spring session을 사용해 redis로 관리.
- 세션 timeout 3600(60분) 설정
### 1-1. 세션 관리 필요성
- 기존에는 세션을 사용하지 않았음. jwt 사용해서 stateless하게 인증을 유지.
- **oauth2 로그인에서 세션을 사용함**: spring security oauth2 로그인에서 defualt AuthorizationRequestRepository는 **HttpSessionOAuth2AuthorizationRequestRepository**를 사용. oauth2 로그인시 인증 요청과 redirect된 요청을 비교해서 인증 요청이 악의적인 공격이 아닌 실제 유저의 요청에 의한 인증이라는 것을 검증하는데 이때 authorizationRequest를 세션에 저장해둠.
- 추후 스케일 아웃을 통해 서버 인스턴스가 여러대가 될 경우 **in-memory 세션 사용시 세션 동기화 문제 발생.**: 기존 인증 요청을 처리한 서버와 oauth2 인증 서버에서 redirect된 요청이 처리되는 서버가 다를 경우 세션에 저장된 authorizationRequest가 없어서 인증 예외 발생.
### 1-2. redis 를 선택한 이유
- 서버 인스턴스 내부 메모리가 아닌 외부 저장소에 세션을 저장해야하는 필요성 대두. 여러 서버가 공통으로 접근할 수 있는 세션 저장소 필요.
- **mysql db가 아닌 redis를 사용하는이유:** 비지니스 db와 세션 db를 분리. redis가 조회가 빠르고 간단. 비지니스 db i/o를 줄임.
- **sticky session이 아닌 redis session을 사용하는 이유**: sticky session 방식은 부하 분산이 제대로 되지 않을 가능성이 있고 서버가 다운될 경우 해당 서버와 관련된 모든 요청 처리가 불가능해질 수 있으므로 위험함.

## 2. 필터체인 분리
- 로그인을 위한 security filter chain과 로그인을 제외한 요청을 위한 securiy filter chain 분리
- 로그인을 위한 필터 (ex) JsonUsernamePasswordAuthenticationFilter, OAuth관련 필터)들을 로그인이 아닌 요청들이 불필요하게 거치지 않도록하기 위함. 1) 성능 문제 2) 디버깅 문제 3) 엔드포인트 관리 쉽도록.
- 로그인을 제외한 요청은 jwt토큰이 필요한 유저 API와, 인증이 필요없는 일반 API로 나뉨. 로그인을 제외한 요청은 spring security에서 자동으로 추가한 필터를 제외하고는 JwtAuthenticaionFilter만 거침.


# ✔️ 해결해야하는 이슈
1. 중복가입을 어떻게 처리할 것인가?
- id/pw로 가입한 사용자가 oauth로 가입하게 된다면?
- 현재는 중복 가입 가능함. 어떤 방식으로 로그인하느냐에 따라서 접근 가능한 유저 정보가 달라짐.
- 추후 중복 가입을 방지하거나, id/pw로 가입한 유저가 oauth로 로그인할 수 있도록 통합할 수 있으면 좋을듯.
2. LoginSuccessHandler 처리 중 발생한 예외를 어떻게 처리할 것이냐. 현재 서블릿 컨테이너의 기본 예외로 처리됨.

i960107 committed:
rename: CustomDaoAuthenticationProvider 삭제

- CustomDaoAuthenticaionProvider: 인증 성공시 CustomAuthenticedPrincipal을 반환하기 위해 만들어두었던 클래스 삭제. CustomAuthenticatedPrincipal을 반환할 필요가 없어짐. DaoAuthenticationProvider 로 대체.

i960107 committed:
feat: user 테이블 oauth 관련 컬럼 추가

- OauthProvider, OauthUserId 컬럼 추가.
- password  nullable하도록 변경.(oauth 가입 유저는 password가 null)

i960107 committed:
feat: oauth 유저 조회 및 유저

- findOauthUser: oauthprovider, oauth userid로 가입된 유저 있는지 조회.
- oauthjoin: oauth로 인증된 유저 저장.

i960107 committed:
feat: oauth provider 네이버 추가

- oauth 프로필 활성화되도록 application.yaml에 추가

i960107 committed:
feat: 인증을 위한 security filter chain 추가

- 인증을 제외한 요청은 securityFilterChain을 거치고, 인증(usernamepassword, oauth ) loginSecurityFilterChain을 거침.
- 인증 제외한 로직이 불필요한 filter를 거치지 않도록하고 디버깅이 쉽도록 함.
- loginSecurityFilterChain을 적용할 Url을 명시하고 securityFilterChain보다 먼저 실행되도록 order를 지정함으로써 인증 Url 구분.
- oauth2Login은 성공시 oauth2AuthenticationSuccessHandler에 의해서 처리되고 실패시 loginFailureHandler에 의해서 처리됨
- username password authentication filter는 username, password를 json 으로 받는 JsonUsernamePasswordAuthenticationFilter를 사용
- username password 로그인은 성공시 usernamePasswordAuthenticationSuccessHandler에 의해서 처리되고, 실패시 loginFailureHandler에 의해 처리됨

i960107 committed:
feat: JsonUserNamePasswordFilter 추가

- username, password 으로 로그인
- 인증 성공시 jwt 토큰을 생성해서 LoginResponse객체를 응답 바디에 담아 보냄.
- 인증 실패시 401 Unauthorized 응답을 보냄. Oauth 로그인과 공통

i960107 committed:
feat: Oauth2Login 기능 추가

- OAuth 로그인 성공시 OAuth2AuthenticationSuccessHandler에서 같은 정보로 등록된 유저가 없다면 유저 정보를 테이블에 저장하고 jwt 토큰을 생성해서 응답
- OAuth2UserAttributeConverter를 사용해 OAuth2User에 담긴 attribute들을  OAuth2UserInfo로 변환. Oauth Provider 확장이 쉽도록 함.

green-ribbon-admin commented on app-api/src/main/java/com/aireview/review/login/oauth/OAuth2UserAttributeConverter.java:
```diff
@@ -0,0 +1,47 @@
+        }
+    };
+    // TODO: 5/9/24 KAKAO 추가 필요
+    String oauthProvider;
```
p3. enum으로 관리하면 좋을거같네요

i960107 replied:
oauthProvider를 타입으로 관리하기 위해서 enum으로 수정했습니다😀

green-ribbon-admin commented on app-api/src/main/java/com/aireview/review/login/usernamepassword/UsernamePasswordAuthenticationSuccessHandler.java:
```diff
@@ -0,0 +1,43 @@
+    public void onAuthenticationSuccess(HttpServletRequest request, HttpServletResponse response, Authentication authentication) throws IOException, ServletException {
+        CustomUserDetails user = (CustomUserDetails) authentication.getDetails();
+
+        String token = jwt.createJwt(user.getUserId(), user.getUsername(), user.getAuthorities());
```
p3.   소셜 로그인과 id/pw가 기준이 다른 이유가 다로 있나요? 
user.getId 와 user.getUserId가 다른걸까요?

String token = jwt.createJwt(user.getId(), user.getEmail(), oauth2User.getAuthorities());
String token = jwt.createJwt(user.getUserId(), user.getUsername(), user.getAuthorities());

i960107 replied:
소셜 로그인과 id/pw의 authentication의 principal 타입이 달라서 해당하는 정보를 가져올때 다른 메서드를 사용했어요
소셜 로그인시에는 OAuth2User가 반환되고 id/pw는 UserDetails가 반환되어요. OAuth2User와 UserDetails 인터페이스를 모두 구현하는 객체를 만들지 않고 OAuth2User와 UserDetails 구현체를 분리 했기때문에 저런 코드가 나왔어요.
또, 해당 코드는 oauth로그인과 id/pw 로그인 성공시의 공통 로직을 추상화하는 방식으로 리팩터링도 해봤어요.
OAuth2User와 UserDetails를 분리한 이유와 추상화한 방식에 대해서 pr 메시지 2-1. LoginSuccessHandler에 추가하였습니다 😄

i960107 committed:
fix: Jpa auditing listener 비활성화 문제 해결

- configuration 어노테이션 추가. AuditingBeanDefinitionRegistrationSupport 가 호출되어서 auditing handler 빈으로 등록되도록 함.

i960107 committed:
fix: 유저 가입 기능 인증 필요없도록 필터 체인 수정

i960107 committed:
fix: AccountController를 통한 로그인 deprecate

-  필터에서 인증과 관련된 모든 수행하도록 수정

i960107 committed:
feat: Username, Password 로그인 필터 구현

- username, password를 통한 로그인 필터 구현. jacksonHttpMessageConverter 사용해서 로그인 요청 데이터 파싱한 후 DaoAuthenticationProvider가 UsernamePasswordAuthenticationToken을 이용한 인증 진행. 인증 완료 후 토큰 반환. 인증 실패시 401 에러 반환.
- CustomUserDetails 생성자 validation 로직 추가.
- 기존 컨트롤러를 통한 로그인에서 사용하던 CustomDaoAuthenticationProvider, LoginRequest ,LoginResponse 클래스 Deprecate.

i960107 committed:
fix: 응답 content type 수정

- application -> application/json으로 수정

i960107 committed:
feat:  redis 저장소 추가

- build.gradle 의존성 추가. spring-data-redis 및 spring-session-redis
- docker-compose redis 컨테이너 추가
- 어플리케이션 설정 추가: redis datasource추가 및 세션 저장소 redis로 지정
- 세션저장소를 redis로 지정한 이유: spring security에서 oauth 로그인시 요청을 검증하기 위해서 세션을 사용하는데 이때 서버 인스턴스가 여러대고 각자 세션을 인메모리에 저장한다면 리다이렉트된 요청이 검증 실패하는 경우가 생김.

i960107 committed:
feat: kakao oauth provider 추가

- application.yaml에 kakao provider 설정 값 추가
- converter에 추가 : kakao에서 가져온 사용자 정보를 OAuth2UserInfo 형식으로 변환. 여러 provider 의 다른 응답형식을 통일
- OAuth2UesrInfo 생성자에 validation 추가.

i960107 committed:
- jwt -> jwtService로 이름 변경
- jwt dto 추가. accessToken 및 refreshToken 포함.
- jwtService 에서 토큰 생성을 위한 createJwt, refreshJwt만 공개(public). accessToken발급시 무조건 RefreshToken도 갱신되도록 강제하기 위함.
- createJwt: 로그인 후 토큰 발급 -> accessToken과 refreshToken 발급 후 redis의 refreshTokenRepository에 저장(key:userId, value:token string). 기존에 해당 유저의 발급된 refreshToken이 있는 경우에 덮어씀
- refreshToken: refreshToken으로 새로운 accessToken 및 refreshToken 발급.  기존에 발급된 refreshToken이 없거나 요청 refreshToken과 다르다면 예외 발생. refreshToken이 다른경우 refreshToken이 탈취된 것으로 간주하고 해당 refreshToken 제거.
- refreshToken 만료 시간은 application.yaml로 부터 읽어온 값으로 설정. refreshToken 만료 시간이 되거나, 잘못된 refresh token으로 토큰 갱신 요청시 refreshToken은 제거됨.
- refreshToken 요청 endpoint 생성 및 인증 없이 요청 가능하도록 filter chain 수정

i960107 committed:
feat: oauth, id/pw 인증 완료시 refreshToken도 반환 및 successHandler 리팩터링

- LoginResponse dto 수정
- LoginSuccessHandler에 공통 로직 추가 및 기존 Oauth2AuthenticationSuccessHandler, UsernamePasswordAuthenticationSuccessHandler가 LoginSuccessHandler를 상속받도록 추가.
- LoginSuccessHandler: processAuthenticationSuccess()라는 추상 메서드를 호출하여 User객체를 얻고 Jwt 를 생성한 후 response에 담기. 상속받는 객체들은 processAuthenticationSuccess()에서 authentication으로부터 User객체를 가져오고, 추가적인 로직을 수행함.
- OAuth2AuthenticationSuccessHandler: OAuth2UserAttributeConverter를 통해 각 provider들의 응답을 OAuth2UserInfo로 변환. 인증된 OAuth2User로 등록된 사용자가 없다면 테이블에 저장 후 User객체 반환.
- UsernamePasswordAuthenticationSuccessHandler: UserDetails 구현체에 User객체 자체를 담은 후 인증 결과 Authentication에서 User 복원.

i960107 committed:
feat: RuntimeException에 대한 exception handler 추가

- runtime exception 발생시 stacktrace가 응답으로 전달되는 것을 방지하기 위해서 exceptionHandler추가. 에러 메시지만 전달.

i960107 committed:
rename: CustomUserDetails 수정

파일 이름 수정

green-ribbon-admin commented on app-api/src/main/java/com/aireview/review/config/RedisConfig.java:
```diff
@@ -0,0 +1,17 @@
+@Configuration
+public class RedisConfig {
+    @Bean
+    public RedisTemplate<String, String> redisTemplate(RedisConnectionFactory redisConnectionFactory) {
```
p3. lettuce 라는거 들어보셧을까요? https://jojoldu.tistory.com/418 을 한번 봐주시면 좋을거같아요 resttemplate으로 가능하지만 성능측면에서 변경하면 좋을거같아요 실제 현업에서는 lettuce를 써용

i960107 replied:
네 jedis에 비해서 lettuce가 스레드 세이프하고, 비동기적으로 여러 연결을 동시에 처리하는 성능이 높고, 클러스터를 효율적으로 제공하기 때문에 jedis보다 많이 사용된다고 이해하였습니다. spring-data-redis를 사용하면 기본적으로 LettuceConnectionFactory를 빈으로 등록하고 lettuce 연결을 사용한다는 것을 확인하였는데, 혹시 다른 부분을 말씀하시는걸까요??😶

green-ribbon-admin commented on app-api/src/main/java/com/aireview/review/login/LoginSuccessHandler.java:
```diff
@@ -0,0 +1,52 @@
+            response.getWriter().flush();
+            response.getWriter().close();
+        } catch (IOException exception) {
+            // TODO: 5/9/24  oauth user 정보가 없으면? IllegalArgumentException 어떻게 처리?
```
p3.확인해봐야할게 LoginFailureHandler 로 빠져서 굳이 이렇게 안해도 될거같긴한데요 
음 그리고 프론트가 없으니.. 약간 애매하네요 어디론가 로그인 성공시 리다이렉션 시켜주는게 일반적인데

green-ribbon-admin commented on app-api/src/main/java/com/aireview/review/login/Role.java:
```diff
@@ -1,7 +1,8 @@
 
 public enum Role {
-    USER("ROLE_USER");
+    USER("ROLE_USER"),
```
p5. 약간 애매한게 이게 유저 타입으로 enum이 되면 좋게 권한하고는 약간 무관한거같아요 
권한에 대해서는 admin, user 이런씩으로 일반유저와 관리자 같은걸로 나누는게 나을거같아요. 
Role 보다는 user 가 로그인 타입에 대해서 저장하는거 같아서

green-ribbon-admin commented on app-api/src/main/java/com/aireview/review/login/usernamepassword/CustomDaoAuthenticationProvider.java:
```diff
@@ -1,19 +1,20 @@
 import org.springframework.security.core.Authentication;
 import org.springframework.security.core.userdetails.UserDetails;
 
+@Deprecated(forRemoval = true)
```
p3. 이거 넣은거는 해당 클래스가 사라질건가요?

i960107 replied:
네 😄 로그인 로직이 변경되면서 CustomDaoAuthneticaitonProvider는 더 이상 필요 없고 DaoAuthenticationProvider를 사용하게 되어서 해당 클래스는 사용되지 않도록 @Deprecated 어노테이션을 추가하였습니다.
